### PR TITLE
MultiProcessing&Actor Implementation

### PR DESCRIPTION
**Pull request by hongdami**
Add Actor feature

Signed-off-by: Miao, Hongda <hongda.miao@intel.com>

Issue Number: 

Objective of pull request:

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/posix_actor.cc [hunk 1]
```diff
@@ -20,14 +23,52 @@ int PosixActor::Stop() {
...
 
   // Check the status
   return 0;
 
 }
 
-int PosixActor::Wait() {
+int PosixActor::Stop() {
+  int status;
+  kill(pid_, SIGTERM);
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/posix_actor.cc:35]: we need to send `stop` cmd/signal to target actor, not kill.
maybe rename as `ForceStop`.

File: .github/workflows/ci-build.yml [hunk 2]
```diff
@@ -150,3 +150,13 @@ jobs:
         .venv\Scripts\activate.ps1
         python3 -m pip install ipykernel
         pytest
+
+    - name: Run message infrastructure unit tests
+      if: runner.os == 'Linux' || runner.os == 'macOS'
```
Comment (killight98) [inline .github/workflows/ci-build.yml:155]: at current, `Linux` only

File: .github/workflows/ci-build.yml [hunk 3]
```diff
@@ -150,3 +150,13 @@ jobs:
         .venv\Scripts\activate.ps1
         python3 -m pip install ipykernel
         pytest
+
+    - name: Run message infrastructure unit tests
+      if: runner.os == 'Linux' || runner.os == 'macOS'
+      run: |
+        git lfs fetch
+        git lfs pull
```
Comment (killight98) [inline .github/workflows/ci-build.yml:158]: don't need lfs?

**Review by killight98 — CHANGES_REQUESTED** [review]

File: src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/message_infrastructure_py_wrapper.cc [hunk 4]
```diff
@@ -41,12 +42,11 @@ PYBIND11_MODULE(MessageInfrastructurePywrapper, m) {
...
   py::class_<SharedMemory> (m, "SharedMemory")
-    .def(py::init<>());
+    .def(py::init<int, int>());
   py::class_<PosixActor> (m, "Actor")
     .def("wait", &PosixActor::Wait)
-    .def("stop", &PosixActor::Stop)
-    .def("get_status", &PosixActor::GetStatus);
-    // .def("trace", &PosixActor::Trace);
+    .def("get_status", &PosixActor::GetActorStatus)
+    .def("stop", &PosixActor::Stop);
```
Comment (seowwj) [inline src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/message_infrastructure_py_wrapper.cc:49]: Do we need to add `force_stop` from `PosixActor::ForceStop`?

File: src/lava/magma/runtime/message_infrastructure/CMakeLists.txt [hunk 5]
```diff
@@ -3,23 +3,43 @@ project(message_passing)
...
-target_link_libraries(MessageInfrastructurePywrapper PRIVATE multi_processing)
+if(PY_WRAPPER)
+    target_include_directories(message_infrastructure PUBLIC 
+                               ${pybind11_INCLUDE_DIRS}
+                               ${PYTHON_INCLUDE_DIRS}
+                               ${PROJECT_SOURCE_DIR}/message_infrastructure/csrc
+                               )
+    exec_program(${PYTHON_EXECUTABLE}
+                 ARGS "-c \"import numpy; print(numpy.get_include())\""
+                 OUTPUT_VARIABLE NUMPY_INCLUDE_DIR
```
Comment (killight98) [inline src/lava/magma/runtime/message_infrastructure/CMakeLists.txt:35]: u need to add `NUMPY_INCLUDE_DIR` into line 28 `target_include_directories`.
i think we could add getting `NUMPY_INCLUDE_DIR` at setup.py.

**Review by killight98 — APPROVED** [review]

**Pull request merged**